astropy/nddata/decorators.py

```diff
@@ -3,161 +3,236 @@
-
+            ignored = []
             if not unpack and isinstance(data, NDData):
                 raise TypeError("Only NDData sub-classes that inherit from {0}"
-                                " can be used by this function".format(accepts.__name__))
+                                " can be used by this function"
+                                "".format(accepts.__name__))
 
-            # If data is an NDData instance, we can try and find properties that
-            # can be passed as kwargs.
+            # If data is an NDData instance, we can try and find properties
+            # that can be passed as kwargs.
             if unpack:
-
-                ignored = []
-
                 # We loop over a list of pre-defined properties
-                for prop in supported_properties:
-
-                    # We only need to do something if the property exists on the
-                    # NDData object
-                    if hasattr(data, prop):
+                for prop in islice(SUPPORTED_PROPERTIES, 1, None):
+                    # We only need to do something if the property exists on
+                    # the NDData object
+                    try:
                         value = getattr(data, prop)
-                        if (prop == 'meta' and len(value) > 0) or (prop != 'meta' and value is not None):
-                            if prop in func_kwargs:
-                                if prop in kwargs and kwargs[prop] is not None:
-                                    warnings.warn("Property {0} has been passed explicitly and as an "
-                                                  "NDData property, using explicitly specified value".format(prop),
-                                                  AstropyUserWarning)
-                                else:
-                                    kwargs[prop] = value
-                            else:
-                                ignored.append(prop)
+                    except AttributeError:
+                        continue
+                    # Skip if the property exists but is None or empty.
+                    if prop == 'meta' and not value:
+                        continue
+                    elif value is None:
+                        continue
+                    # Warn if the property is set but not used by the function.
+                    propmatch = attr_arg_map.get(prop, prop)
+                    if propmatch not in func_kwargs:
+                        ignored.append(prop)
+                        continue
+
+                    # Check if the property was explicitly given and issue a
+                    # Warning if it is.
+                    if propmatch in kwargs:
+                        # If it's in the func_args it's trivial but if it was
+                        # in the func_kwargs we need to compare it to the
+                        # default.
+                        # FIXME: This obviously fails if the explicitly given
```

Comment:
Just to make absolutely sure I understand the issue: it is that a keyword argument to the function will always have a value (this is what makes it a keyword), so you are trying to check whether the value is different than the default and only issue the warning if they are different.

But that case is indistinguishable from the case where the user has explicitly given a value that is the default.

In that case does an unnecessary warning generated or does the property from the `NDData` object get used instead? Looks to me like the latter, which seems more problematic.